### PR TITLE
fix #492 - adding link to github

### LINKED ISSUE #492: Dereference data models Entity type to the corresponding JSON Schema
For those URIs in the @context that point to Entity types such as 

https://uri.fiware.org/ns/datamodels/Building 

should be resolved to the corresponding JSON Schema, in that particular case 

https://fiware.github.io/dataModels/specs/Building/Building/schema.json

We need to do that for all the Schemas we already have in the Data Models

/cc @jason-fox

### PR DESCRIPTION
**Pull request by caa06d9c**
URLs can be verified by executing

`cat context.jsonld | python -m json.tool | grep github | awk '{ print $2 }' | sed -e 's/,//g' | xargs -I "{}" -L 1 curl -sSq -I {} | grep HTTP`

Everywhere 200

### CODE REVIEW

**Comment by jmcanterafonseca** [comment]
be careful, we don't want to change the URIs of the Entity types, we want just a redirection to happen.

**Comment by caa06d9c** [comment]
So as an example, WeatherObserved, `https://uri.fiware.org/ns/datamodels/entity-types/WeatherObserved` should return 301 to `https://fiware.github.io/dataModels/specs/Weather/WeatherObserved/schema.json` ?

File: tools/ldcontext_generator.py [hunk 1]
```diff
@@ -1,23 +1,53 @@
...
+# Here a list of spec alerts will be stored
+alert_list = [
+]
+
+# Template to prepare a valid URL of a schema for a mapping
+schema_url = 'https://github.com/FIWARE/dataModels/blob/master/{}'
+
+# Agri* schemas stores at another github organization
+agri_url = 'https://github.com/GSMADeveloper/NGSI-LD-Entities/blob/master/definitions/{}.md'
+agri_mapping = {
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: can we externalize these mappings into a file?

File: tools/ldcontext_generator.py [hunk 2]
```diff
@@ -173,6 +209,31 @@ def aggregate_ld_context(f, uri_prefix, predefined_mappings):
     for p in ld_context:
         aggregated_context[p] = ld_context[p]
 
+        if p not in mapping_list['terms']:
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: can we put this code that has just to do with the yaml generation in a separate function?

File: tools/ldcontext_generator.py (hunk 1 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Done

File: tools/ldcontext_generator.py (hunk 2 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Almost done, not sure if in a correct form

File: tools/ldcontext_generator.py [hunk 3]
```diff
@@ -1,23 +1,45 @@
...
+ - context.jsonld, that serves https://schema.lab.fiware.org/ld/fiware-datamodels-context.jsonld
+ - mapping_list.yml, that serves  https://uri.fiware.org/ns/datamodels
+
+context.jsonld is combined by extracting the properties, types and  enumerations of a JSON Schema and
+converting them into terms of a JSON-LD @Context. mapping_list.yml uses the result of extracting
+to prepare a list of terms with schemas and specifications.
 
 Copyright (c) 2019 FIWARE Foundation e.V.
 
 Author: José M. Cantera
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: add yourself also as author

File: tools/ldcontext_generator.py [hunk 4]
```diff
@@ -1,23 +1,45 @@
...
 
 # Here the aggregated @context will be stored
 aggregated_context = {
 }
 
+# Here a list of mappings (terms->schemas/specifications) will be stored
+terms_list = {
+    "terms": {}
+}
+
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: can you ellaborate what an alert is?

File: tools/ldcontext_generator.py [hunk 5]
```diff
@@ -149,22 +177,22 @@ def schema_2_ld_context(schema, uri_prefix, predefined_mappings):
...
     elif os.path.isdir(input_file):
         for f in (os.listdir(input_file)):
             process_file(os.path.join(input_file, f),
-                         uri_prefix, predefined_mappings)
+                         uri_prefix, predefined_mappings, terms_mappings)
 
 
-def aggregate_ld_context(f, uri_prefix, predefined_mappings):
+def aggregate_ld_context(f, uri_prefix, predefined_mappings, terms_mappings):
     global aggregated_context
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: we need to also declare as global terms_list, don't we?

File: tools/ldcontext_generator.py [hunk 6]
```diff
@@ -173,6 +201,39 @@ def aggregate_ld_context(f, uri_prefix, predefined_mappings):
...
+
+        terms_list['terms'][p]['schemas'].append(schema_url.format(f.split('../')[1]))
+
+        file_to_add = find_file(f, terms_mappings)
+        if file_to_add:
+            terms_list['terms'][p]['specifications'].append(file_to_add)
+        else:
+            alert_list.append(f)
+
+
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: can you add comment describing what this function does?

File: tools/ldcontext_terms_mappings.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+    "AgriCrop": "Agri-Crop",
+    "AgriGreenhouse": "Agri-Greenhouse",
+    "AgriParcelOperation": "Agri-Parcel-Operation",
+    "AgriParcelRecord": "Agri-Parcel-Record",
+    "AgriParcel": "Agri-Parcel",
+    "AgriPest": "Agri-Pest",
+    "AgriProductType": "Agri-Product-Type"
+}
```
Comment (jmcanterafonseca) [inline tools/ldcontext_terms_mappings.json]: nit: new line needed

File: tools/ldcontext_generator.py (hunk 4 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Yes, if associated with the schema spec file is not exists, the script will inform about it

File: tools/ldcontext_generator.py [hunk 8]
```diff
@@ -1,23 +1,45 @@
...
 # Here the aggregated @context will be stored
 aggregated_context = {
 }
 
+# Here a list of mappings (terms->schemas/specifications) will be stored
+terms_list = {
+    "terms": {}
+}
+
+# Here a list of terms alerts will be stored
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: elaborate more this comment

File: tools/ldcontext_generator.py [hunk 9]
```diff
@@ -1,23 +1,45 @@
...
+alert_list = [
+]
+
+# Template to prepare a valid URL of a schema for a term mapping
+schema_url = 'https://fiware.github.io/dataModels/{}'
+specification_url = 'https://github.com/FIWARE/dataModels/blob/master/{}'
+
+# Agri* schemas stores at another github organization
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: stored

**Comment by jmcanterafonseca** [comment]
please ensure autopep8 is passed

thanks

File: tools/ldcontext_generator.py (hunk 4 above)
Comment (jmcanterafonseca) [reply tools/ldcontext_generator.py]: please add this explanation to the comment ;)

File: tools/ldcontext_generator.py (hunk 3 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

File: tools/ldcontext_generator.py (hunk 4 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

File: tools/ldcontext_generator.py (hunk 5 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

File: tools/ldcontext_generator.py (hunk 6 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

File: tools/ldcontext_terms_mappings.json (hunk 7 above)
Comment (caa06d9c) [reply tools/ldcontext_terms_mappings.json]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

File: tools/ldcontext_generator.py (hunk 8 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

File: tools/ldcontext_generator.py (hunk 9 above)
Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed in 8933b1ef27836a637fd56c3530e5b226b30a2fdb

**Comment by caa06d9c** [comment]
> please ensure autopep8 is passed
> 
> thanks

Fixed

File: tools/ldcontext_generator.py [hunk 10]
```diff
@@ -1,23 +1,46 @@
...
 aggregated_context = {
 }
 
+# The list of mappings (term->schema/specification) will be stored here
+terms_list = {
+    "terms": {}
+}
+
+# The list of terms alerts will be stored here (if the specification file
+# associated with the term doesn't exists)
```
Comment (jmcanterafonseca) [inline tools/ldcontext_generator.py]: nit exist

Comment (caa06d9c) [reply tools/ldcontext_generator.py]: Fixed

**Comment by jmcanterafonseca** [comment]
LGTM, good work!

**Pull request merged**